integrations/access/common/app.go

```diff
@@ -172,6 +188,148 @@ func (a *BaseApp) onWatcherEvent(ctx context.Context, event types.Event) error {
+	process.OnTerminate(func(ctx context.Context) error {
+		cancel()
+		return nil
+	})
+
+	s := gocron.NewScheduler(time.UTC).Cron(accessListReviewCron)
+	log := logger.Get(ctx)
+
+	log.Info("Access list monitor is running")
+
+	job, err := s.Do(func() {
+		log.Info("Looking for Access List Review reminders")
+
+		var nextToken string
+		var err error
+		for {
+			var accessLists []*accesslist.AccessList
+			accessLists, nextToken, err = a.accessLists.ListAccessLists(ctx, 0 /* default page size */, nextToken)
+			if err != nil {
+				log.Errorf("error listing access lists: %v", err)
+				return
+			}
+
+			for _, accessList := range accessLists {
+				if err := a.notifyForAccessListReviews(ctx, accessList); err != nil {
+					log.WithError(err).Warn("Error notifying for access list reviews")
+				}
+			}
+
+			if nextToken == "" {
+				break
+			}
+		}
+	})
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	// Start the scheduler.
+	s.StartAsync()
+	a.alMonitorJob.SetReady(true)
+
+	select {
+	case <-ctx.Done():
+		s.Stop()
+	case <-job.Context().Done():
+		return trace.Wrap(job.Error())
+	}
+
+	log.Info("Access list monitor is finished")
+
+	return nil
+}
+
+// notifyForAccessListReviews will notify if access list review dates are getting close. At the moment, this
+// only supports notifying owners.
+func (a *BaseApp) notifyForAccessListReviews(ctx context.Context, accessList *accesslist.AccessList) error {
+	log := logger.Get(ctx)
+	allRecipients := make(map[string]Recipient, len(accessList.Spec.Owners))
+
+	now := a.clock.Now()
+	// Find the current notification window.
+	notificationStart := accessList.Spec.Audit.NextAuditDate
+
+	// If the current time before the notification start time, skip notifications.
+	if now.Before(notificationStart) {
+		log.Infof("Access list %s is not ready for notifications, notifications start at %s", accessList.GetName(), notificationStart.Format(time.RFC3339))
+		return nil
+	}
+
+	// Get the owners from the bot as recipients.
+	for _, owner := range accessList.Spec.Owners {
+		recipient, err := a.bot.FetchRecipient(ctx, owner.Name)
+		if err != nil {
+			log.WithError(err).Debugf("error getting recipient %s", owner.Name)
+			continue
+		}
+		allRecipients[owner.Name] = *recipient
+	}
+
+	if len(allRecipients) == 0 {
+		return trace.NotFound("no recipients could be fetched")
+	}
+
+	// Try to create base notification data with a zero notification date. If these objects already
+	// exist, that's okay.
+	for _, recipient := range allRecipients {
+		owner := recipient.Name
+		_, err := a.accessListPluginData.Create(ctx, accessListOwnerKey(accessList.GetName(), owner), pd.AccessListNotificationData{
+			User:             owner,
+			LastNotification: time.Time{},
+		})
+
+		// Error is okay so long as it's already exists.
+		if err != nil && !trace.IsAlreadyExists(err) {
+			return trace.Wrap(err, "error during create")
+		}
+	}
+
+	windowStart := notificationStart.Add(now.Sub(notificationStart) / oneWeek)
```

Comment:
Can't say I fully follow what's being done here by dividing the duration by one week 🤔

Reply:
I've corrected this, but the intent was to divide by the number of weeks, get the week count, and then add it back in to effectively get the next window for initiating notifications.